### DISCUSSION TITLE
Lesson 12 : Broken Import Statement

### DISCUSSION THREAD

**Discussion by SquilliamX** [Cyfrin/foundry-full-course-cu#804] [Q&A]
Hi everyone, I am on lesson 12 but it seems i have a broken Import statement in my OracleLib.sol file. When i import the AggregatorV3Interface, I receive an error saying ```Source "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol" not found: File not found. Searched the following locations: "".solidity(6275)```. I recieve this error even though i copied and pasted the import statement from DSCEngine.sol, where it works perfect. If I move my OracleLib.sol out of the library folder in src and just leave it in src with no library folder, the error goes away and the code works fine. However if I leave OracleLib.sol in libraries and then change my import statement to ``` import {AggregatorV3Interface} from
    "lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol"; ``` the error goes away, but a new one appears. When I try to ctrl+click into AggregatorV3Interface.sol, it then says "The file is not displayed in the text editior because it is a directory." Does anyone know how I can fix my import statement and keep OracleLib.sol in my libraries folder in src? Thanks!

```
remappings = [
    "@chainlink/contracts=lib/chainlink-brownie-contracts/contracts/",
    '@openzeppelin/contracts=lib/openzeppelin-contracts/contracts',
]
```

OracleLib.sol - 
```
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

/**
 * @title OracleLib
 * @author SquilliamX
 * @notice This library is used to check the Chainlink Oracle for stale data.
 * If a price is stale, the function will revert, and render the DSCEngine unusable - this is by design
 * We want the DSCEngine to freeze if prices become Stale.
 *
 * So if the Chainlink network explodes and you have money in the protocol... too bad
 */
library OracleLib {
    function StalePriceCheck(AggregatorV3Interface priceFeed)
        public
        view
        returns (uint80, uint256, uint256, uint256, uint80)
    {
        priceFeed.latestRoundData();
    }
}

```

**Comment by theirrationalone** [comment]
Hey @SquilliamX !

The Problem is `/` appended at the end on the right hand side into the remapping of `@chainlink/contracts`.

look at this:
```toml
remappings = [
    "@chainlink/contracts=lib/chainlink-brownie-contracts/contracts/", # here at the end `/` is the problem 😉 which is wrong ❌
    '@openzeppelin/contracts=lib/openzeppelin-contracts/contracts',
]
```

correct is ✅:
```toml
remappings = [
    "@chainlink/contracts=lib/chainlink-brownie-contracts/contracts", 
    '@openzeppelin/contracts=lib/openzeppelin-contracts/contracts',
]
```

**Reply by SquilliamX** [reply]
Hey thanks for the reply, I tried this and it didn't solve my issue. Do you have any other ideas? thanks

**Reply by theirrationalone** [reply]
Dear @SquilliamX !

have you tried to quote openzeppelin's remapping with double `"` like this:

```toml
remappings = ["@chainlink/contracts=lib/chainlink-brownie-contracts/contracts", "@openzeppelin/contracts=lib/openzeppelin-contracts/contracts",]
```

**Reply by SquilliamX** [reply]
yes i have, it doesn't change anything. It is my @chainlink import that breaks when i move my OracleLib.sol into my libraries folder inside my src folder. However the import statement works when I just keep Oraclelib.sol in src folder and not in libraries folder.

**Reply by theirrationalone** [reply]
Make sure you have this config in your `foundry.toml`...

```toml
[profile.default]
src = "src"
out = "out"
libs = ["lib"]
remappings = ['@chainlink/contracts/=lib/chainlink-brownie-contracts/contracts/', '@openzeppelin/contracts=lib/openzeppelin-contracts/contracts']
```

If you have this config or if not, Please update.
After this Please try to restart your vs-code and then try again.

**Reply by SquilliamX** [reply]
yes i do indeed have that, and my import statement is still broken :(.
 
My foundry.toml looks like this
```
[profile.default]
src = "src"
out = "out"
libs = ["lib"]
remappings = [
    "@chainlink/contracts=lib/chainlink-brownie-contracts/contracts",
    '@openzeppelin/contracts=lib/openzeppelin-contracts/contracts',
]

[invariant]
runs = 128
depth = 128
fail_on_revert = true

```

**Reply by theirrationalone** [reply]
try to do this... 

make a file named as remappings.txt

Open it and paste below into it. 

```txt
@chainlink/contracts=lib/chainlink-brownie-contracts/contracts
 chainlink-brownie-contracts/=lib/chainlink-brownie-contracts/
@openzeppelin/contracts=lib/openzeppelin-contracts/contracts
 ds-test/=lib/forge-std/lib/ds-test/src/ 
 forge-std/=lib/forge-std/src/ 
 foundry-devops/=lib/foundry-devops/
```

Then try again.

**Comment by sunnyStefi** [comment]
@SquilliamX is the AggregatorV3Interface working well inside DSCEngine? Could you provide us with the whole OracleLib location and code? I cannot reproduce your issue. Thanks!
